src/utils/DeleteSet.js

```diff
@@ -278,37 +278,13 @@ export const readAndApplyDeleteSet = (decoder, transaction, store) => {
     const state = getState(store, client)
     for (let i = 0; i < numberOfDeletes; i++) {
       const clock = decoder.readDsClock()
-      const clockEnd = clock + decoder.readDsLen()
+      const len = decoder.readDsLen()
+      const clockEnd = clock + len
       if (clock < state) {
         if (state < clockEnd) {
           addToDeleteSet(unappliedDS, client, state, clockEnd - state)
         }
-        let index = findIndexSS(structs, clock)
-        /**
-         * We can ignore the case of GC and Delete structs, because we are going to skip them
-         * @type {Item}
-         */
-        // @ts-ignore
-        let struct = structs[index]
-        // split the first item if necessary
-        if (!struct.deleted && struct.id.clock < clock) {
-          structs.splice(index + 1, 0, splitItem(transaction, struct, clock - struct.id.clock))
-          index++ // increase we now want to use the next struct
-        }
-        while (index < structs.length) {
-          // @ts-ignore
-          struct = structs[index++]
-          if (struct.id.clock < clockEnd) {
-            if (!struct.deleted) {
-              if (clockEnd < struct.id.clock + struct.length) {
-                structs.splice(index, 0, splitItem(transaction, struct, clockEnd - struct.id.clock))
-              }
-              struct.delete(transaction)
-            }
-          } else {
-            break
-          }
-        }
+        applyDeleteItem(transaction, structs, { clock, len })
```

Comment:
This may sound silly, but creating a function call for every deleted `Item` really hurts performance.

Yjs is able to handle millions of `Item` integrations because the "hot-path" is inlined. I realize that there is more code that could be inlined. 

Please don't create a separate function call for deleting items.

Reply:
Hm, actually, I always thought that modern JS engines do function inlining itselves, so I didn't worry about that too much. With your updated approach using encoding/decoding we won't need `applyDeleteItem`, so I will remove this function.

Reply:
You are right. JavaScript engines do inline code on a predictive basis. In nodejs you can inspect how many functions "can" be inlined and how many functions are actually inlined on execution. It seems there is an inline-budget (because inlining does duplicate code and increases the size of the heap). So inlining functions beforehand, if possible, does improve performance. I just started a few weeks ago to work on inlining the hot-path myself to see some slight improvements of 10-20%. I seriously don't recommend doing that for all projects. But it made sense for Yjs.